index.ts

```diff
@@ -0,0 +1,536 @@
+            },
+        };
+        this.hooks = {
+            "after:deploy:deploy": this.hookWrapper.bind(this, this.setupBasePathMapping),
+            "after:info:info": this.hookWrapper.bind(this, this.domainSummary),
+            "before:remove:remove": this.hookWrapper.bind(this, this.removeBasePathMapping),
+            "create_domain:create": this.hookWrapper.bind(this, this.createDomain),
+            "delete_domain:delete": this.hookWrapper.bind(this, this.deleteDomain),
+        };
+    }
+
+    /**
+     * Wrapper for lifecycle function, initializes variables and checks if enabled.
+     * @param lifecycleFunc lifecycle function that actually does desired action
+     */
+    public async hookWrapper(lifecycleFunc: any) {
+        this.initializeVariables();
+        if (!this.enabled) {
+            this.serverless.cli.log("serverless-domain-manager: Custom domain is disabled.");
+            return;
+        } else {
+            return await lifecycleFunc.call(this);
+        }
+    }
+
+    /**
+     * Lifecycle function to create a domain
+     * Wraps creating a domain and resource record set
+     */
+    public async createDomain(): Promise<void> {
+        const certArn = await this.getCertArn();
+        const domainInfo = await this.createCustomDomain(certArn);
+        await this.changeResourceRecordSet("UPSERT", domainInfo);
+        this.serverless.cli.log(
+            `Custom domain ${this.givenDomainName} was created/updated.
+            New domains may take up to 40 minutes to be initialized.`,
+        );
+    }
+
+    /**
+     * Lifecycle function to delete a domain
+     * Wraps deleting a domain and resource record set
+     */
+    public async deleteDomain(): Promise<void> {
+        let domainInfo;
+        try {
+            domainInfo = await this.getDomainInfo();
+        } catch (err) {
+            if (err.message === `Error: ${this.givenDomainName} not found.`) {
+                this.serverless.cli.log(`Unable to delete custom domain ${this.givenDomainName}.`);
+                return;
+            }
+            throw err;
+        }
+        await this.deleteCustomDomain();
+        await this.changeResourceRecordSet("DELETE", domainInfo);
+        this.serverless.cli.log(`Custom domain ${this.givenDomainName} was deleted.`);
+    }
+
+    /**
+     * Lifecycle function to create basepath mapping
+     * Wraps creation of basepath mapping and adds domain name info as output to cloudformation stack
+     */
+    public async setupBasePathMapping(): Promise<void> {
+        await this.createBasePathMapping();
+        const domainInfo = await this.getDomainInfo();
+        this.addOutputs(domainInfo);
+        await this.printDomainSummary(domainInfo);
+    }
+
+    /**
+     * Lifecycle function to delete basepath mapping
+     * Wraps deletion of basepath mapping
+     */
+    public async removeBasePathMapping(): Promise<void> {
```

Comment:
Not sure if we need a wrapper function here since its a 1 liner. Could put thee contents of `deleteBasePathMapping` in here

Reply:
I like the idea of making the lifecycle functions and the actual API calls separate - just consistent with the rest of the codebase.